Add ability to return copy of the rotated vector

dvdvideo1234 opened the pull request.

Kefta commented:
I would remove the nil check line, it will just lead to people not knowing when they sent in a bad value. Also, you can skip the Vector.Set line by using the Vector's copy-constructor. Ex. ``local v = Vector(self)``

dvdvideo1234 commented:
If modifiers returned `*this` the constructor will become:

function meta:GetRotated(ang)
	return Vector(self):Rotate(ang)
end

thegrb93 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,164 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Returns a copy of bisector of two vectors
+-----------------------------------------------------------]]
+function meta:Bisect(vec)
```
Typo

Kefta replied:
Where?

thegrb93 replied:
Look 10 lines up

dvdvideo1234 replied:
Ten lines up is
```
--[[---------------------------------------------------------
Convets the vector to bisector of two vectors
-----------------------------------------------------------]]
function meta:Bisect(vec)
```

fruitwasp replied:
`GetBisect` instead.

fruitwasp replied:
GetBisect

dvdvideo1234 replied:
It seems correct o me: https://www.mathsisfun.com/geometry/bisect.html

thegrb93 replied:
It's still not fixed, dummy

robotboy655 commented:
I feel like most of the stuff in here is useless and there are several problems with other stuff:
1) Do we really need stuff like `ToTable`, `ToAngle` (this one specially confusing because the name might make someone think its transforming a normalized/direction vector into an angle, which it does not), `GetSub/Mul/Add/Div`? `GetZero`? `Middle`?
2) You still have 2 Vector.Bisect methods
3) Your first vector.Bisect sets global variables.
4) Shouldn't ```GetNegate``` be ```GetNegated```? ```Normalize``` and `GetNormlized`
5) `GetMiddle` never uses its argument, and probably should be `GetCenter` or something
6) `GetResize` also never uses its argument, and should be `GetResized`

dvdvideo1234 commented:
1. The idea is to change the data structure, but if you insist, I will remove them ;)
2. Thanks, fixed.
3. Thanks, fixed also
4. Fixed ;)
5. The `Middle` must use its argument for obtaining a [midpoint](https://en.wikipedia.org/wiki/Midpoint).
6. Thanks, also fixed ;)

Kefta commented:
SetEx should be added in the engine, not Lua. I requested it here https://github.com/Facepunch/garrysmod-requests/issues/515

Kefta commented:
I think these should be implemented in the engine for speed.

Kefta commented:
You should utilise vector:Unpack where you index x, y, and z to be more efficient. Also, using number indexing is slightly faster than string indexing (vec[1] vs vec.x).

dvdvideo1234 commented:
> You should utilise vector:Unpack

Last time I tested at home Unpack was not pushed to production.

> Also, using number indexing is slightly faster than string indexing (vec[1] vs vec.x)

Yes, this is true. [The official wiki says we must use X, Y, Z](Also, using number indexing is slightly faster than string indexing (vec[1] vs vec.x)). It is always better to stick to the documentation. ;)

Kefta commented:
That is an oversight of the wiki, I will change it. You should still use numerical indexing.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,176 @@ local meta = FindMetaTable( "Vector" )
+Convets the vector to nagated version of itself
+-----------------------------------------------------------]]
+function meta:Negate()
+	self.x = -self.x
```
Unpack here.

dvdvideo1234 replied:
Done. Added single component trigger. Id only one is needed to be negated. When no arguments, all are negated.

Kefta commented:
Updated the wiki for the Vector and Angle pages on all possible ways to index them.

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,176 @@ local meta = FindMetaTable( "Vector" )
 -----------------------------------------------------------]]
-function meta:ToColor( )
+function meta:ToColor()
+	return Color(self.x * 255, self.y * 255, self.z * 255)
```
`Color(x * 255, y * 255, z * 255)`

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,176 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Converts Vector to Table - removes all metrods from the copy
+-----------------------------------------------------------]]
+function meta:ToTable(arr)
```
Done.

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,176 @@ local meta = FindMetaTable( "Vector" )
+Converts Vector to Angle - data is copied
+-----------------------------------------------------------]]
+function meta:ToAngle()
+	return Angle(self.x, self.y, self.z)
```
Done.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Converts Vector to Table - removes all metrods from the copy
+-----------------------------------------------------------]]
+function meta:ToTable(arr)
```
This argument is no longer needed

dvdvideo1234 replied:
Done !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+-----------------------------------------------------------]]
+function meta:ToArray()
+	local x, y, z = self:Unpack()
+	return {x, y, z}
```
``return {self:Unpack()}``

dvdvideo1234 replied:
Done !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+-----------------------------------------------------------]]
+function meta:ToAngle()
+	local x, y, z = self:Unpack()
+	return Angle(x, y, z)
```
``return Angle(self:Unpack())``

dvdvideo1234 replied:
Done !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+Adds a unpacked x, y and z to a vector
+-----------------------------------------------------------]]
+function meta:AddEx(x, y, z)
+	self[1] = self[1] + x
```
I think ``*Ex`` should be changed to ``*Unpacked`` for naming consistency. Also use ``Unpack`` and ``SetUnpacked`` here.

dvdvideo1234 replied:
Ok !

dvdvideo1234 replied:
Done with Add/Sub*

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+	self[1] = self[1] + x
+	self[2] = self[2] + y
+	self[3] = self[3] + z
+	return self
```
I wouldn't return here for consistency with other vector modifying methods.

dvdvideo1234 replied:
`Bisect` is a modifier also. As I remember `a: Add(b)` on vectors does not return anything. Sorry. This due to my personal coding style for returning `*this` in C++.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Convets the vector to nagated version of itself
+-----------------------------------------------------------]]
+function meta:Negate(bx, by, bz)
```
I think these arguments are pointless. Negate should negate the entire vector and not just parts - that can be done inline in code if necessary instead of inducing a bunch of type checks.

dvdvideo1234 replied:
Ok. Thanks !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,247 @@ local meta = FindMetaTable( "Vector" )
+	local x = self:Dot(ang:Forward())
+	local y = self:Dot(ang:Right())
+	local z = self:Dot(ang:Up())
+	self[1], self[2], self[3] = x, y, z
```
``SetUnpacked`` here.

dvdvideo1234 replied:
Done !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Convets the vector to nagated version of itself
+-----------------------------------------------------------]]
+function meta:Nudge(num, vec)
```
What are the use cases for this function? Where does the term "nudge" from?

I think it would just be better to write this inline as the multiplied vector will probably be reused in most cases.

dvdvideo1234 replied:
A short term for `origin = origin + coefficient * direction`

Kefta replied:
Ah, so this is the same as [VectorMA](https://github.com/VSES/SourceEngine2007/blob/43a5c90a5ada1e69ca044595383be67f40b33c61/src_main/public/mathlib/mathlib.h#L269). Like Negate, this should be binded from the engine then instead of Lua.

dvdvideo1234 replied:
Ah, never new it has to be a dedicated engine function. I use it mery much in my addons, So I thought it must be needed here ;) Thank you !

dvdvideo1234 replied:
I will Remove Nudge. This is the exact thing I am talking about.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:Project(vec)
```
It would be better to copy the original vector instead of using GetNormalized than setting it after finding the dot product.
```Lua
local v = Vector(self)
self:Normalize()
self:Mul(v:Dot(self))
```

dvdvideo1234 replied:
No, because it need Normalized of the argument as it defines the base X coordinate axis. You have to store the argumant in `self` rather.

Kefta replied:
Yeah you're correct, mixed those up.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Retrieves the angle between two vectors from 0 to math.pi
+-----------------------------------------------------------]]
+function GetAngle(vec)
```
Missing ``meta:`` in front. This should probably renamed to AngleBetween to prevent confusion with vec:Angle(). Also, this could be more efficiently implemented as

```Lua
return math.acos(self:Dot(vec) / math.sqrt(self:LengthSqr() * vec:LengthSqr()))
```

using the formula ``cos(a) = dot(x, y)/(|x|*|y|)``. While you're at it, you should also add an additional optional argument to specify a normal vector representing the plane the vectors lie in. This can be written as

```Lua
return math.atan2(self:Determinant(vec, normal:GetNormalized()), self:Dot(vec))
```

using the formula ``tan(a) = determinant(x, y, n) / dot(x, y)``. Most library functions in GMod don't use type checking so you can get away with

```Lua
function meta:AngleBetween(vec, n)
	if (n == nil) then
		return ...
	end

	return ...
end
```

dvdvideo1234 replied:
That sounds neat. Thank you !

dvdvideo1234 replied:
Never noticed the second parameter but it sounds great ;)

dvdvideo1234 replied:
Done !

```
local a = Vector(3,3)
local b = Vector(3,0)
local c = Vector(0,0,6)
local d = a:GetProject(b)
print(d, a:AngleBetween(b,c))
```

dvdvideo1234 replied:
VEC{3,0,0}	-0.78539816339745

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Calculates 3x3 determinant by using other two vectors as rows
+-----------------------------------------------------------]]
+function meta:GetDeterminant(vec2, vec3)
```
I would just make the name ``Determinant`` to be consistent with ``Vector:Length``.

dvdvideo1234 replied:
Ok. Fixed.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
-function meta:ToColor( )
+function meta:ToColor(num)
+	local x, y, z = self:Unpack()
+	return Color(x * 255, y * 255, z * 255, (num or 0) * 255)
```
This should be ``num == nil and 1 or num`` (the ``== nil`` to prevent people accidentally sending in ``false`` as the argument and never noticing the difference, but more importantly having 1 as the default so the alpha defaults to 255).

dvdvideo1234 replied:
Yes, thank you. Fixed ;)

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,237 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Convets the vector to nagated version of itself
+-----------------------------------------------------------]]
+function meta:Negate()
```
Ok.

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:GetProject(...)
```
This must modify `self` this is different than `GetProject` which makes a copy. ;)

dvdvideo1234 replied:
This is done for better support where `A:getSomething` utilizes `A:Someting` and returns a copy ;)

dvdvideo1234 replied:
Written that way you only need to change the code in `meta:Project(...)` to adjust the behavior.

Kefta replied:
It may be less duplicitous to have ``Get*`` call down to the non-returning alternatives, but it's slower. I guess it's not that big of a deal anyway.

Kefta replied:
It may be less duplicitous to have Get* call down to the non-returning alternatives, but it's slower. I guess it's not that big of a deal anyway.

dvdvideo1234 replied:
Yes. The performance difference is negligible and you only need to change one method in the future instead of both. If `Project` works, then `GetProject` works too ;)

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,237 @@ local meta = FindMetaTable( "Vector" )
 
-	return Color( self.x * 255, self.y * 255, self.z * 255 )
+--[[---------------------------------------------------------
+Convets the vector to nagated version of itself
```
Yes, thanks ! Fixed.

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,237 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Adds a unpacked x, y and z to a vector
+-----------------------------------------------------------]]
+function meta:AddUnpacked(x, y, z)
```
Thank you. Fixed :)

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
 -----------------------------------------------------------]]
-function meta:ToColor( )
+function meta:ToColor(num)
+	local a = (num == nil and 1 or num)
```
This doesn't need to be a local - it can be placed inline in the Color call.

dvdvideo1234 replied:
Ok. Done.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+-----------------------------------------------------------]]
+function meta:AddUnpacked(x, y, z)
+	local vx, vy, vz = self:Unpack()
+	self:AddUnpacked(vx + x, vy + y, vz + z)
```
I think you meant to do SetUnpacked here, this is infinite recursion right now.

dvdvideo1234 replied:
Yes ! Thanks !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+-----------------------------------------------------------]]
+function meta:SubUnpacked(x, y, z)
+	local vx, vy, vz = self:Unpack()
+	self:SubUnpacked(vx - x, vy - y, vz - z)
```
Same here.

dvdvideo1234 replied:
Same. Yes. Done !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Returns a copy of the nudged vector
+-----------------------------------------------------------]]
+function meta:GetNudge(...)
```
This needs to be removed if Nudge is.

dvdvideo1234 replied:
Done !

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Convets the vector to negated version of itself
+-----------------------------------------------------------]]
+function meta:Negate()
```
This should be removed for an engine implementation - I would go ahead and also remove GetNegated until the bindings are made so this can be used on people's servers in the meantime.

dvdvideo1234 replied:
Removed for you to do engine implementation.

Kefta commented:
Looking good. Last thing to clean up is the style (spaces around parentheses and and brackets, C-styled logical operators) to match the codebase, and to change the name of the PR as this is no longer one method.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,230 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:GetProject(...)
```
Shouldn't this be GetProjected to match GetBisected and GetResized?

dvdvideo1234 replied:
Yes, it should. Fixed.

Kefta replied:
It's not

dvdvideo1234 replied:
Ah.... Noticed the code bas is `Add` -> `GetAdd`

dvdvideo1234 replied:
Reverted. Also Added ray intersection fraction calculator

dvdvideo1234 replied:
On the other hand should I rename `GetAdd` `GetSub` to `GetAdded` `GetSubbed` .... :confused:

Kefta replied:
I think all Get methods except for Midpoint and Add, Sub Mul, Div should end with ``ed`` since we currently have ``Unpack`` and ``SetUnpacked``. So ``GetProjected``, ``GetBisected``, ``GetResized``, and ``GetRotated``.

dvdvideo1234 replied:
Ah. Makes sense. Thank you! Updated !

Kefta replied:
Whoops, forgot to mention to leave Basis alone, as well.

dvdvideo1234 replied:
Ok, Reverted! Thanks.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+	f1 --> Intersection fraction of the first ray
+	f2 --> Intersection fraction of the second ray
+-----------------------------------------------------------]]
+function meta:IntersectFraction( vD1, vO2, vD2 )
```
You have a bug with:
```Lua
local dn = ( dx:Length() ) ^ 2
if( dn <= 0 ) then return nil end
```
since x^2 will always be >= 0.

I don't recognise this formula - what is the "intersection fraction"? Since this takes two independent vectors and directions, I think this would be better suited for the ``math`` or ``util`` libraries.

dvdvideo1234 replied:
Should I remove it then ?

Kefta replied:
Can you link to a mathematical/code source from where you got this formula from?

dvdvideo1234 replied:
Em, I know it from the university ... Maths part 2 analytical geometry. But I am sure that you can find it in the web also ;)

dvdvideo1234 replied:
It can be done with `== 0`, though I like <= better :D. Anywayn` is actually `|dx|^2` the length of the cross procuct of the directions. Maybe it is not needed in the vector library .Removed ;)

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:GetBasised( ... )
```
"Basised" isn't a word, should still be ``GetBasis``.

dvdvideo1234 replied:
Fixed. Thanks !

dvdvideo1234 commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+Retrieves the angle between two vectors from 0 to math.pi
+-----------------------------------------------------------]]
+function meta:AngleBetween( vec, nrm )
+	if( nrm == nil ) then
```
Done !

Kefta commented:
I think this is finally in a good place. There's some additional optimisations that can be done to save on Vector objects in ``Get*`` alternatives instead of just calling the original functions with a copy, but those can be done post-merge. Thanks for staying patient with all the corrections.

dvdvideo1234 commented:
Thank you too for adding all these ;)

dvdvideo1234 commented:
Thoughts @thegrb93 , @fruitwasp is this any good ?

thegrb93 commented:
The ToAngle needs to be removed as it will be confused with Vector:Angle

Kefta commented:
I think ``<operation>Unpacked`` should be added in the engine - it would be more efficient than creating a vector object simply to perform an operation on all three digits, and even more efficient than setting each component manually. They could even be made into optional args to be able to do two component operations in one function call instead of 4 (2 __newindex, 2 __index). I'll add them to my request on https://github.com/Facepunch/garrysmod-requests/issues/1442.

Now looking at it, ``ToAngle`` should be removed not for the confusion with ``Angle``, but rather the uselessness of the operation: I can think of no cases where you would go from a Vector to Angle unless it were to convert an impulse force, but that would require performing operations on the component anyway.

dvdvideo1234 commented:
The APIs with `Unpacked` will work very nice for me, as I will remove like 100+ lines of code from my addon just for using these. Sometimes a copy of the vector of the executed operation is needed,  and for this `Get<Operation>Unpacked` is used.

Kefta commented on garrysmod/lua/includes/extensions/vector.lua:
```diff
@@ -259,3 +259,23 @@ function meta:AngleBetween( vec, nrm )
+ * bN   > When set to true, normalizes the three unit vectors
+ * vR   > The function returns the right vector on success
+-----------------------------------------------------------]]
+function meta:Orthogonal(vF, bN)
```
I don't think this should be part of the lib - it's arbitrary for the up vector to be the self arg. Maybe it should be a global function or thrown into util.

dvdvideo1234 replied:
Should I remove it then?

Kefta replied:
I mean, it's your PR, it's ultimately up to Rubat. I just think it's vague implemented as a vector class method personally.

dvdvideo1234 replied:
You are probably right ;)

dvdvideo1234 commented:
@Kefta @thegrb93,

Is it better for data conversion to just [run this](https://github.com/Facepunch/garrysmod/blob/0b20ee02bd0558f1841317fdb303fbfb19b83f63/garrysmod/lua/includes/extensions/util.lua#L106). It relies on the object factory to create an instance from any type. I see the conversion from a string is not needed. But I wonder, what is the difference between:

1. `Vector(1,2,3)`  This is the regular call
2. `Vector({1,2,3})` Table-array with 1,2,3
3. `Vector({x=1,y=2,z=3})` Table-hash with X,Y,Z
4. `Vector("1 2 3")` This according to the wiki parses the string to a vector. Notice the space.
5. `Vector("1,2,3")` What will happen when I give it another delimiter. Notice the comma. :P
6. `Vector(true,true,true)` This I think should produce `Vector(1,1,1)` dunno..

I cannot think of any suitable usage converting:
1. `vector` from `function`,
2. `vector` from `userdata`
3. `vector` from a `thread`.

Kefta commented:
2 doesn't work but can be achieved with ``Vector(unpack({1, 2, 3})``. 3, 5, and 6 don't work. You shouldn't adjust your functions at all for that behaviour - let the Vector constructor handle all type coercions. This PR is fine in its current state.

dvdvideo1234 commented:
@Kefta Thanks!

Requested in https://github.com/Facepunch/garrysmod/pull/1619

dvdvideo1234 commented:
@Kefta 
I see that [there is a  new wiki](https://wiki.facepunch.com/gmod/Angle:SetUnpacked). does this mean that the PR #1419 is merged already?

robotboy655 commented:
> @ Kefta
> I see that [there is a new wiki](https://wiki.facepunch.com/gmod/Angle:SetUnpacked). does this mean that the PR #1419 is merged already?

No, I have added some methods that were requested [here](https://github.com/Facepunch/garrysmod-requests/issues/515)

CornerPin commented:
What is the point of having `GetAdd`, `GetSub`, `GetMul` and `GetDiv` when we already have `__add`, `__sub`, `__mul` and `__div`? The special metamethods do the same things and must be more efficient.

dvdvideo1234 commented:
@CornerPin 

These are here for language richness. I can admit, they do the same stuff like the other ones, but I've never measured it.
The richest one language is, the better as you can so the same thing by utilizing different lines of code.
They are here for language richness mostly, for the users who prefer `local c = a:GetAdd(b)` more than `local c = a + b`.

dvdvideo1234 commented:
@robotboy655 

Should I remove the `VECTOR:Get%` part of the PR as @CornerPin sais we can use `+, - * /` ?

I can leave only the methods that return nothing ?
